### ISSUE TITLE
KT-31077 - android.kotlinOptions block is lacking its type

### ISSUE THREAD

**Issue by mannodermaus** [gradle/kotlin-dsl-samples#1368]
[YouTrack Issue](https://youtrack.jetbrains.com/issue/KT-31077)

After applying the `kotlin-android` plugin, a global configuration closure named `kotlinOptions` becomes available inside the `android` block. This closure takes one parameter, however it's not strongly typed and resolved to `Action<Any>`:

```kotlin
public fun com.android.build.gradle.internal.dsl.BaseAppModuleExtension.kotlinOptions(configure: org.gradle.api.Action<kotlin.Any>): kotlin.Unit { /* compiled code */ }
```

### Expected Behavior

I can access the implicit `KotlinJvmOptions` directly:

```kotlin
android {
  kotlinOptions {
    jvmTarget = "1.8"
  }
```

### Current Behavior

The missing typing prevents us from setting the configuration without a proper cast in place:

```kotlin
  kotlinOptions {
    // We have to add the explicit cast before accessing the options itself.
    // If we don't, it does not work: "unresolved reference: jvmTarget"
    val options = this as KotlinJvmOptions
    options.jvmTarget = "1.8"
  }
}
```

Alternatively, we need to configure each task individually:

```kotlin
tasks.withType<KotlinCompile> {
  kotlinOptions {
    options.jvmTarget = "1.8"
  }
}
```

### Your Environment
- Gradle 5.3.1
- Gradle kotlin-dsl 1.1.3
- Android Studio 3.4.0 RC3
- Kotlin Plugin 1.3.21-release-Studio3.4-1

**Comment by gildor** [comment]
I believe this is not a problem of Kotlin DSL, but this convention. Is it part of kotlin-android plugin feature? If so, better to report it on Kotlin issue tracker

**Comment by mannodermaus** [comment]
Thanks for the reply. I wasn't sure where this convention is being synthesized, but I do believe that it's added through the `kotlin-android` plugin. I'll report it on the Kotlin issue tracker and attach a link here.

**Comment by StefMa** [comment]
Can you please refer to the ticket in the Jetbrains issue tracker?! Thank you :)

**Comment by gildor** [comment]
I think that the problem with usage of extension function `addExtenion`. This function add extension without type:
https://github.com/JetBrains/kotlin/blob/b6102c52b9dac6a258a2add4a4a28571a502692c/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/gradleUtils.kt#L52-L53

Instead, it should be something like this:
```
internal inline fun <reified T : Any> Any.addExtension(name: String, extension: T) =
    (this as ExtensionAware).extensions.add(T::class.java, name, extension)
```

So in this case extension will be properly typed.

**Comment by mannodermaus** [comment]
I raised the ticket as [KT-31077](https://youtrack.jetbrains.com/issue/KT-31077) on YouTrack now. Added the link to the OP, too!

**Comment by eskatos** [comment]
Fixed in Kotlin 1.3.60 🎉

**Comment by guuilp** [comment]
@eskatos it was fixed in Gradle 5.6, actually: https://youtrack.jetbrains.com/issue/KT-31435.

I tested here with kotlin 1.3.60 and still got the error. When I updated to Gradle 5.6.x, it worked.
